hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3AFileSystem.java

```diff
@@ -0,0 +1,110 @@
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.regex.Pattern;
+
+/**
+ * Unit tests for {@link S3AFileSystem}.
+ */
+public class TestS3AFileSystem extends Assert {
+  final File TEMP_DIR = new File("target/build/test/TestS3AFileSystem");
+  final String longStr = // 1024 char
+    "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key";
+  final String longStrTruncated = "very_long_s3_key__very_long_s3_key__";
+
+  @Rule
+  public Timeout testTimeout = new Timeout(30 * 1000);
+
+  @Before
+  public void init() throws IOException {
+    Files.createDirectories(TEMP_DIR.toPath());
+  }
+
+  @After
+  public void teardown() throws IOException {
+    File[] testOutputFiles = TEMP_DIR.listFiles();
+    for(File file: testOutputFiles) {
+      Files.delete(file.toPath());
+    }
+    Files.deleteIfExists(TEMP_DIR.toPath());
+  }
+
+  @Before
+  public void nameThread() {
+    Thread.currentThread().setName("JUnit");
+  }
+
+  /**
+   * Test the {@link S3AFileSystem#safeCreateTempFile(String, String, File)}.
+   * The code verifies that the input prefix and suffix don't exceed the file system's max name
+   * length and cause an exception.
+   *
+   * This test verifies the basic contract of the process.
+   */
+  @Test
+  public void testSafeCreateTempFile() throws Throwable {
+    // fitting name isn't changed
+    File noChangesRequired = S3AFileSystem.safeCreateTempFile("noChangesRequired", ".tmp", TEMP_DIR);
+    assertTrue(noChangesRequired.exists());
+    String noChangesRequiredName = noChangesRequired.getName();
+    assertTrue(noChangesRequiredName.startsWith("noChangesRequired"));
+    assertTrue(noChangesRequiredName.endsWith(".tmp"));
+
+    // a long prefix should be truncated
+    File excessivelyLongPrefix = S3AFileSystem.safeCreateTempFile(longStr, ".tmp", TEMP_DIR);
```

Comment:
each of these should be split into their own test method so they can fail independently. I know, we ignore that in ITests, but that is because they're integration tests with longer setup overhead